File: ClientRuntimes/NodeJS/ClientRuntime.Azure/lib/userTokenCredentials.js
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information. 

var util = require('util');
var msrest = require('ms-rest');
var adal = require('adal-node');
var Constants = msrest.Constants;
var AzureEnvironment = require('./azureEnvironment');

/**
* Creates a new SubscriptionCredentials object.
*
* @constructor
* @param {string} 
* @param {string} authorizationScheme The authorization scheme.
*/
function UserTokenCredentials(clientId, domain, username, password, clientRedirectUri, options) {
  if (Boolean(clientId) || typeof clientId !== 'string') {
    throw new Error('clientId must be a non empty string.');
  }
  
  if (Boolean(domain) || typeof domain !== 'string') {
    throw new Error('domain must be a non empty string.');
  }
  
  if (Boolean(username) || typeof username !== 'string') {
    throw new Error('username must be a non empty string.');
  }
  
  if (Boolean(password) || typeof password !== 'string') {
    throw new Error('password must be a non empty string.');
  }
  
  if (Boolean(clientRedirectUri) || typeof clientRedirectUri !== 'string') {
    throw new Error('clientRedirectUri cannot be null.');
  }
  
  if (!options.environment) {
    this.environment = AzureEnvironment.Azure;
  } else {
    this.environment = options.environment;
  }
  
  if (!options.authorizationScheme) {
    this.authorizationScheme = 'bearer';
  } else {
    this.authorizationScheme = options.authorizationScheme;
  }
  
  this.clientId = clientId;
  this.domain = domain;
  this.username = username;
  this.password = password;
  this.clientRedirectUri = clientRedirectUri;
}

util.inherits(UserTokenCredentials, msrest.TokenCredentials);

/**
* Signs a request with the Authentication header.
*
* @param {webResource} The WebResource to be signed.
* @param {function(error)}  callback  The callback function.
* @return {undefined}
*/
UserTokenCredentials.prototype.signRequest = function (webResource, callback) {
  var authorityUrl = this.environment.authenticationEndpoint + '/' + this.domain;
  var context = new adal.AuthenticationContext(authorityUrl, this.environment.validateAuthority /*,exports.tokenCache*/);
  
  context.acquireToken(this.environment.tokenAudience, this.username, this.clientId, function (err, result) {
    if (err) {
      return callback(new Error('Failed to acquire token. \n' + err));
    }
    
    webResource.headers[Constants.HeaderConstants.AUTHORIZATION] = 
      util.format('%s %s', this.authorizationScheme, result.accessToken);
    callback(null);
  });
};

module.exports = UserTokenCredentials;

File: src/client/Java/azure-client-authentication/src/main/java/com/microsoft/azure/credentials/UserTokenCredentials.java
/**
 *
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License. See License.txt in the project root for license information.
 *
 */

package com.microsoft.azure.credentials;

import com.microsoft.aad.adal4j.AuthenticationContext;
import com.microsoft.aad.adal4j.AuthenticationResult;
import com.microsoft.azure.AzureEnvironment;
import com.microsoft.rest.credentials.TokenCredentials;

import java.io.IOException;
import java.util.Date;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

/**
 * Token based credentials for use with a REST Service Client.
 */
public class UserTokenCredentials extends TokenCredentials {
    /** The Active Directory application client id. */
    private String clientId;
    /** The domain or tenant id containing this application. */
    private String domain;
    /** The user name for the Organization Id account. */
    private String username;
    /** The password for the Organization Id account. */
    private String password;
    /** The Uri where the user will be redirected after authenticating with AD. */
    private String clientRedirectUri;
    /** The Azure environment to authenticate with. */
    private AzureEnvironment environment;
    /** The current authentication result. */
    private AuthenticationResult authenticationResult;

    /**
     * Initializes a new instance of the UserTokenCredentials.
     *
     * @param clientId the active directory application client id.
     * @param domain the domain or tenant id containing this application.
     * @param username the user name for the Organization Id account.
     * @param password the password for the Organization Id account.
     * @param clientRedirectUri the Uri where the user will be redirected after authenticating with AD.
     * @param environment the Azure environment to authenticate with.
     *                    If null is provided, AzureEnvironment.AZURE will be used.
     */
    public UserTokenCredentials(String clientId, String domain, String username, String password, String clientRedirectUri, AzureEnvironment environment) {
        super(null, null); // defer token acquisition
        this.clientId = clientId;
        this.domain = domain;
        this.username = username;
        this.password = password;
        this.clientRedirectUri = clientRedirectUri;
        if (environment == null) {
            this.environment = AzureEnvironment.AZURE;
        } else {
            this.environment = environment;
        }
    }

    /**
     * Gets the active directory application client id.
     *
     * @return the active directory application client id.
     */
    public String getClientId() {
        return clientId;
    }

    /**
     * Gets the tenant or domain the containing the application.
     *
     * @return the tenant or domain the containing the application.
     */
    public String getDomain() {
        return domain;
    }

    /**
     * Gets the user name for the Organization Id account.
     *
     * @return the user name.
     */
    public String getUsername() {
        return username;
    }

    /**
     * Gets the password for the Organization Id account.
     *
     * @return the password.
     */
    public String getPassword() {
        return password;
    }

    /**
     * Gets the Uri where the user will be redirected after authenticating with AD.
     *
     * @return the redirecting Uri.
     */
    public String getClientRedirectUri() {
        return clientRedirectUri;
    }

    /**
     * Gets the Azure environment to authenticate with.
     *
     * @return the Azure environment to authenticate with.
     */
    public AzureEnvironment getEnvironment() {
        return environment;
    }

    @Override
    public String getToken() throws IOException {
        if (authenticationResult != null
            && authenticationResult.getExpiresOnDate().before(new Date())) {
                acquireAccessTokenFromRefreshToken();
        } else {
            acquireAccessToken();
        }
        return authenticationResult.getAccessToken();
    }

    @Override
    public void refreshToken() throws IOException {
        acquireAccessToken();
    }

    private void acquireAccessToken() throws IOException {
        String authorityUrl = this.getEnvironment().getAuthenticationEndpoint() + this.getDomain();
        AuthenticationContext context = new AuthenticationContext(authorityUrl, this.getEnvironment().isValidateAuthority(), Executors.newSingleThreadExecutor());
        try {
            authenticationResult = context.acquireToken(
                    this.getEnvironment().getTokenAudience(),
                    this.getClientId(),
                    this.getUsername(),
                    this.getPassword(),
                    null).get();
        } catch (Exception e) {
            throw new IOException(e.getMessage(), e);
        }
    }

    private void acquireAccessTokenFromRefreshToken() throws IOException {
        String authorityUrl = this.getEnvironment().getAuthenticationEndpoint() + this.getDomain();
        ExecutorService executor = Executors.newSingleThreadExecutor();
        AuthenticationContext context = new AuthenticationContext(authorityUrl, this.getEnvironment().isValidateAuthority(), executor);
        try {
            authenticationResult = context.acquireTokenByRefreshToken(
                    authenticationResult.getRefreshToken(),
                    this.getClientId(),
                    null, null).get();
        } catch (Exception e) {
            throw new IOException(e.getMessage(), e);
        } finally {
            executor.shutdown();
        }
    }
}
